### ISSUE TITLE
Much slower debugger whith conditional break points

### ISSUE THREAD

**Issue by MkazemAkhgary** [dotnet/roslyn#24888]
I'm not sure if this is the right place. 

I'm wondering why breakpoints with condition are much slower than normal breakpoints or no breakpoints.

If I put no breakpoint and run my program it takes about 1 second to finish. If I put conditional break point it takes 5 seconds to reach that break point. 

 obviously doing nothing is faster than doing something. but a simple condition like `p >= 757000` shouldn't cause that much performance hit.

Is this normal?

**Comment by CyrusNajmabadi** [comment]
> I'm wondering why breakpoints with condition are much slower than normal breakpoints or no breakpoints.

A conditional breakpoint still causes the program to break each time it is hit.  Then the debugger has to go and evaluate the state of the program against the condition you specified.  That condition needs to be go through the entire debugger evaluation system, which involves coordinating between the debugger (shared across languages) and specific language providers (i assume C# in this case).  Then, that expression has to actually *evaluate* (so now you're including the runtime).   There's a lot of machinery and moving parts to make this happen.

**Comment by MkazemAkhgary** [comment]
@CyrusNajmabadi  Hmm, makes sense but.. cant this be done better? 

I know I know. it doesn't sound easy to describe that process, let alone optimizing it! but It would be great. 

since I have nothing to propose about how to optimize this process Ill close this issue.

**Comment by MkazemAkhgary** [comment]
@CyrusNajmabadi 

Oh, just recalled something!

I find putting if statement in actual code and putting simple breakpoint (not conditional) will be a lot faster. as fast as running the program I say.

Cant debugger change the compiled code in runtime? I guess simple conditions that can be written at compile time can be optimized in this way. for example

```
int x = something;  // put conditional breakpoint "x > 100"
```

change the compiled code into

```
int x = something;
if(x > 100) Debugger.Break();
```

If changing compiled code isn't easy process either (with current runtime capabilities) ill close this proposal.

**Comment by svick** [comment]
> If changing the compiled code in runtime isn't easy process either (with current runtime capabilities) ill close this proposal.

Isn't that what Edit and Continue does? Couldn't that machinery be used for this?

Even if it didn't work in all cases (since Edit and Continue has its limitations and since debugger expressions are not the same as C# expressions), I think that making conditional breakpoints significantly faster in the common case could be worth it.

**Comment by jinujoseph** [comment]
cc @tmat

**Comment by tmat** [comment]
Yes, it could do EnC in some cases. As a workaround you can also do that yourself and change the code while debugging instead of adding a conditional break-point. The benefit of the debugger doing it automatically is that you wouldn't need to remember to undo the change once you are done debugging.
